### PR TITLE
942 check that passwords match um password requirements

### PR DESCRIPTION
**Pull request by Eashana0104**
Closes https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/issues/942

The UM password requirements are as follows:

- Must be between 10 and 30 characters in length.

- Must contain at least three of the following character types:

1. Uppercase letters
2. Lowercase letters
3. Numbers
4. Special characters: _ # % ( ) . ^ { } !

- Must differ from the previous five passwords.

- Must not contain the user’s name or UMNetID.

To enforce these rules, I created four custom password validators:

1. LengthRangeValidator – Ensures the password length is within the specified range.

2. CharacterCategoriesValidator – Validates that the password contains the required mix of character types.

3. PasswordHistoryValidator – Ensures the new password is not the same as any of the last five passwords.

4. UserAttributeContainsValidator – Prevents the password from containing the user’s name or UMNetID.

The first three were straightforward to implement, but the PasswordHistoryValidator required a dedicated model. I created a PasswordHistory model to store the last five passwords for each user, rather than adding this directly to the User model, in keeping with the principle of separation of concerns.

With this approach, every time a user changes their password or a new user is created, the hashed password is stored in their password history.

I also wrote comprehensive end-to-end tests in the ttest_custom_validators_e2e.py file to ensure all validators work as intended and handle edge cases effectively.

### CODE REVIEW

**Comment by Eashana0104** [comment]
@danloveg While working on this issue and writing some E2E tests, I noticed that the UserAttributeSimilarityValidator ("NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator") isn’t behaving as expected — it’s not detecting when the name is directly included in the password.

To address this, I increased its specificity, but that change causes it to ignore other errors and always show the similarity error, even when it’s not relevant.

I see two possible approaches:

- Create a custom validator specifically for this case.
- Limit the similarity check to only compare against email addresses.

**Comment by danloveg** [comment]
> @danloveg While working on this issue and writing some E2E tests, I noticed that the UserAttributeSimilarityValidator ("NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator") isn’t behaving as expected — it’s not detecting when the name is directly included in the password.

If you read the code, can you tell where this behaviour is happening or why its not behaving as you expect it to? https://github.com/django/django/blob/fda3c1712a1eb7b20dfc91e6c9abae32bd64d081/django/contrib/auth/password_validation.py#L163

> To address this, I increased its specificity, but that change causes it to ignore other errors and always show the similarity error, even when it’s not relevant.

Can you clarify what you mean by "increasing its specificity"?

**Comment by Eashana0104** [comment]
> > @danloveg While working on this issue and writing some E2E tests, I noticed that the UserAttributeSimilarityValidator ("NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator") isn’t behaving as expected — it’s not detecting when the name is directly included in the password.
> 
> If you read the code, can you tell where this behaviour is happening or why its not behaving as you expect it to? https://github.com/django/django/blob/fda3c1712a1eb7b20dfc91e6c9abae32bd64d081/django/contrib/auth/password_validation.py#L163
> 
> > To address this, I increased its specificity, but that change causes it to ignore other errors and always show the similarity error, even when it’s not relevant.
> 
> Can you clarify what you mean by "increasing its specificity"?

In Django’s UserAttributeSimilarityValidator, the quick_ratio() function from Python’s difflib.SequenceMatcher is used to get a rough estimate of similarity. This method isn’t fully accurate and can produce false positives, which makes it less reliable in certain cases.

The “specificity” I was referring to is essentially the similarity threshold. By default, Django sets this to 0.7, which is quite lenient — meaning it may miss cases where a name is clearly part of the password. To try and fix that, I lowered the threshold so the comparison would be stricter.

However, this made the validator overly sensitive: even passwords that only shared a couple of letters with the name were being rejected. The results became unpredictable and not in line with our needs.

To resolve this, I implemented a custom validator called UserAttributeContainsValidator. Instead of relying on fuzzy matching, it simply checks if the full attribute values (first name, last name, username) are directly present in the password, and accepts or rejects based on that. This gives us consistent and predictable results.

**Comment by danloveg** [comment]
> For the test_custom_validators_e2e.py tests, I’ve currently hardcoded the strings for the different error messages that appear after form validation.
> 
> Do you think it’s okay to leave them as hardcoded values, or would it be better to assign unique IDs to each error message in the UI and reference those IDs in the tests instead? @danloveg

I actually have a different recommendation - before you add any more tests or change the tests, I'll submit a review for this PR.

I'm ready to review this whenever you're ready, but just an FYI, the tests will need to be re-worked

**Comment by Eashana0104** [comment]
> > For the test_custom_validators_e2e.py tests, I’ve currently hardcoded the strings for the different error messages that appear after form validation.
> > Do you think it’s okay to leave them as hardcoded values, or would it be better to assign unique IDs to each error message in the UI and reference those IDs in the tests instead? @danloveg
> 
> I actually have a different recommendation - before you add any more tests or change the tests, I'll submit a review for this PR.
> 
> I'm ready to review this whenever you're ready, but just an FYI, the tests will need to be re-worked

I see, Yeah for sure, I am open to different and more efficient and robust implementations for testing part! We can review the functionality and fix the tests and docs part after that!

File: app/recordtransfer/validators.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                code="password_not_enough_categories",
+                params={"specials": self.allowed_specials},
+            )
+
+    def get_help_text(self) -> str:
+        """Return help text describing the required character categories for the password."""
+        return _(
+            "Your password must include at least three of the following types: uppercase, lowercase, "
+            "numbers, or one of these special characters: _ # % ( ) . ^ { } !"
+        )
```
Comment (danloveg) [inline app/recordtransfer/validators.py:85]: This should use `self.allowed_specials` instead of a hardcoded list. The reason being, that self.allowed_specials may be different from this list.

```suggestion
        return _(
            "Your password must include at least three of the following types: uppercase, lowercase, "
            "numbers, or one of these special characters: %(specials)s"
        )
        % {
            "specials": " ".join(c for c in self.allowed_specials)
        }
```

File: app/recordtransfer/validators.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        # Then check against password history
+        recent_hashes = PasswordHistory.get_recent_password_hashes(
+            user=user, limit=self.history_depth
+        )
+        if any(check_password(password, old_hash) for old_hash in recent_hashes):
+            raise ValidationError(
+                _("You cannot reuse your previous %(n)d passwords."),
+                code="password_in_history",
+                params={"n": self.history_depth},
+            )
```
Comment (danloveg) [inline app/recordtransfer/validators.py:115]: It's conceivable that a system admin could set `self.history_depth=1`. Because of this, you should use `ngettext_lazy` instead.

```suggestion
            raise ValidationError(
                ngettext_lazy(
                    "You cannot reuse your previous password.",
                    "You cannot reuse your previous %(num)s passwords.",
                    self.history_depth,
                )
                % {
                    "num": self.history_depth,
                },
                code="password_in_history",
            )
```

I didn't recommend this for the password length validator because no sane system admin would set a minimum or maximum password length to only 1!

File: app/recordtransfer/validators.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    def get_help_text(self) -> str:
+        """Return help text describing the password history requirement."""
+        return _("Your password must be different from your previous five passwords.")
```
Comment (danloveg) [inline app/recordtransfer/validators.py:119]: "five" may be inaccurate depending on `self.history_depth`. I recommend parameterizing this instead.

```suggestion
        return ngettext_lazy(
            "Your password must be different from your previous password.",
            "Your password must be different from your previous %(num)s passwords.",
            self.history_depth,
        )
        % {
            "num": self.history_depth,
        }
```

File: app/recordtransfer/validators.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+
+class UserAttributeContainsValidator:
+    """Reject passwords that contain the full first name, last name, or username
+    (case-insensitive). Email is intentionally ignored.
+    """
+
+    def __init__(self, user_attributes: tuple[str, ...] | None = None):
+        # Only check whole attribute values for these fields; do not check email
+        self.user_attributes = user_attributes or ("username", "first_name", "last_name")
```
Comment (danloveg) [inline app/recordtransfer/validators.py:129]: This class intentionally only checks username, first_name, and last_name. But a system admin *could* add `email` to the list of checked attributes. You should rename this class, and disable setting the attribute.

```suggestion
class ContainsUserNameValidator:
    """Reject passwords that contain the full first name, last name, or username
    (case-insensitive). Email is intentionally ignored.
    """

    self.user_attributes: ClassVar[tuple] = ("username", "first_name", "last_name")
```

File: app/app/settings/base.py [hunk 5]
```diff
@@ -100,19 +100,22 @@
...
-        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
+        "NAME": "recordtransfer.validators.PasswordHistoryValidator",
+        "OPTIONS": {"history_depth": 5},
     },
     {
-        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
+        "NAME": "recordtransfer.validators.UserAttributeContainsValidator",
+        "OPTIONS": {
+            "user_attributes": ("username", "first_name", "last_name"),
+        },
```
Comment (danloveg) [inline app/app/settings/base.py:118]: Remove these options here, we'll disable overriding the default attributes since the validator only considers those three atrributes.


```suggestion
```

File: app/recordtransfer/validators.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                    f"This password is too long. It must contain at most {self.max_length} characters."
+                ),
+                code="password_too_long",
+            )
+
+    def get_help_text(self) -> str:
+        """Return help text describing the required password length range."""
+        return _(
+            f"Your password must be between {self.min_length} and {self.max_length} characters long."
+        )
```
[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/validators.py]: Resolved by https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/961/commits/67d607b76ef8206f87e7bd052f54ccbb91a28c37

File: app/recordtransfer/tests/e2e/test_custom_validators_e2e.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        """Shows a success alert when changing password with valid inputs."""
+        self.open_profile_and_wait()
+        self.submit_password_change(
+            current="InitialPassword123!",
+            new="SecondPassword456!",
+            confirm="SecondPassword456!",
+        )
+        WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.CLASS_NAME, "alert-success"))
+        )
```
[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/tests/e2e/test_custom_validators_e2e.py]: Resolved by https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/961/commits/d039883d2fd83398e62f549f25b57bb65a009348

File: app/recordtransfer/validators.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        has_special = any(c in self.allowed_specials for c in pwd)
+        count = sum([has_upper, has_lower, has_digit, has_special])
+        if count < self.required_categories:
+            raise ValidationError(
+                _(
+                    "This password must contain at least three of the following: uppercase, lowercase, "
+                    "numbers, or one of these special characters: %(specials)s"
+                ),
+                code="password_not_enough_categories",
+                params={"specials": self.allowed_specials},
```
[the remark this answers is not in the document]
Comment (danloveg) [reply app/recordtransfer/validators.py]: It's possible that in other other languages, you would need to write this sentence differently depending on the number.

In English, you *could* use the plural sentence and it will make sense no matter the number, but that may not be true in other languages.

It's best to use `ngettext_lazy` here for that reason so that translators have as much freedom to translate this into the proper format

[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/validators.py]: I see! This is very well thought of, I ll switch it to ngettext_lazy, Thanks for helping me understand as to why this was necessary!

File: app/recordtransfer/tests/e2e/test_custom_validators_e2e.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+            new="InitialPassword123!",
+            confirm="InitialPassword123!",
+        )
+        WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.CLASS_NAME, "text-error"))
+        )
+        page_text = self.driver.page_source
+        assert "Your new password cannot be the same as your current password." in page_text
+
+    def test_too_short_password_shows_error(self) -> None:
```
[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/tests/e2e/test_custom_validators_e2e.py]: I have added quite a few unit tests here! Added as much as I could think of! Resolved by https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/961/commits/49693ebdda89e9f0f86fec1d15c4b5fb45305058

**Review by danloveg — CHANGES_REQUESTED** [review]
This is just about good to go.

I just pushed a commit that shows an oversight - there are other forms where the user's password can change, but those forms don't save a PasswordHistory for the user.

Instead of adding a PasswordHistory in forms, can you use a `pre_save` signal so that any time the user is saved with a new password, their old password hash is saved? That way you don't have to remember to add this logic to a form so there's no chance of a user's password history not being saved.

File: app/recordtransfer/static/recordtransfer/css/base/tooltip.css [hunk 10]
```diff
@@ -11,6 +11,7 @@
     /* Fade in/out */
     opacity: 0;
     transition: opacity 0.2s ease-in-out;
+    transform: translateX(-12px);
```
[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/static/recordtransfer/css/base/tooltip.css:14]: It was a minor error which i thought this fixed, I ll probably create another issue for this after and revert this change

[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/static/recordtransfer/css/base/tooltip.css]: Resolved by https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/961/commits/ee99df6c8cca0a71564ac3e23c375fd9ae11e588

File: app/recordtransfer/forms/user_forms.py [hunk 11]
```diff
@@ -199,6 +189,20 @@ def clean(self) -> dict[str, Any]:
...
+                    self.add_error("new_password", e)
+
+            if (
+                self.instance.pk
+                and current_password
+                and self.instance.check_password(current_password)
+                and new_password
+                and new_password != current_password
+            ):
+                PasswordHistory.objects.create(user=self.instance, password=self.instance.password)
```
[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/forms/user_forms.py]: To prevent any further oversights with this, I’ve moved the PasswordHistory creation into a pre_save signal in models.py. This ensures it’s triggered consistently whenever a password is changed, regardless of the form used.

Additionally, I noticed that for custom forms, password validation needs to be explicitly invoked, our other custom forms don’t change passwords, so this hasn’t come up before. Django’s built-in forms already call our validate function automatically, but custom ones require adding that call manually.

File: app/recordtransfer/forms/user_forms.py [hunk 12]
```diff
@@ -109,7 +109,6 @@ class UserAccountInfoForm(forms.ModelForm):
         widget=forms.PasswordInput(attrs={"id": HtmlIds.ID_NEW_PASSWORD}),
         label=_("New Password"),
         required=False,
-        validators=[password_validation.validate_password],
```
[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/forms/user_forms.py:112]: We could definitely do that! My main concern is that if we implement it as you suggest, the errors from self._validate_password_change() wouldn’t appear until password_validation.validate_password() passes without any errors. This could be problematic, as we would likely want the user to be notified immediately if they enter their current password incorrectly.

File: app/recordtransfer/models.py [hunk 13]
```diff
@@ -1484,3 +1497,28 @@ def update_upon_save(
     if instance.upload_session:
         instance.upload_session.touch()
         instance.reset_reminder_email_sent()
+
+
+@receiver(pre_save, sender=User)
+def _handle_password_change(instance: User, **kwargs):
```
[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/models.py]: This makes sense to me! Resolved by https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/961/commits/61d67bc1257fd68fadbd09a28c765a42b7948a3f

File: app/recordtransfer/forms/user_forms.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (danloveg) [reply app/recordtransfer/forms/user_forms.py:112]: I'm still a little bit confused, can you give me a scenario for this issue?

[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/forms/user_forms.py:112]: For sure, lets assume we reverted the changes to user_forms.py. 
Screenshot 2025-08-15 at 11 25 30 AM

Currently, if I enter an incorrect current password and an invalid new password, two different functions handle these validations. Since Django automatically validates the form, it calls validate_password to check the new password first. I believe the expected behavior should be for the user to be notified immediately if they entered the wrong current password, rather than waiting until the new password is validated.

[the remark this answers is not in the document]
Comment (Eashana0104) [reply app/recordtransfer/forms/user_forms.py]: Another issue I think could be, if there are multiple errors during validation of new password, all of them should show, It might be frustrating for the user to not be able to see all the requirements beforehand

**Comment by Eashana0104** [comment]
Do you think it would make sense to display all validation errors at once, or is the current approach fine as it is?

Right now, if the new password is invalid, only the first error is shown.

Pros of showing all errors at once:

- Users can see everything they need to fix in a single attempt, avoiding the back-and-forth of fixing one issue only to discover another afterward.
- It can reduce frustration and speed up the process, especially for forms with multiple fields or rules.

Cons of showing all errors at once:

- It may feel overwhelming if there are too many issues listed at once.

- Focusing on one error at a time can keep the form cleaner and help guide the user step-by-step.

**Comment by danloveg** [comment]
> @danloveg Do you think it would make sense to display all validation errors at once, or is the current approach fine as it is?

I think it's less overwhelming to have only one error show. I think it's good as it is

**Comment by danloveg** [comment]
> I think this PR should be good to go then!

Perfect, thank you for being open to feedback, I think this PR is much cleaner and well tested now!

**Review by danloveg — APPROVED** [review]
